### PR TITLE
Changes in SVG backend to improve compatibility with Affinity designer

### PR DESCRIPTION
**Pull request by LorenzoPeri17**
Separate font attributes && use translate(x y) in RendererSVG._draw_text_as_path

## PR summary

Closes [#20910](https://github.com/matplotlib/matplotlib/issues/20910)

Two small changes in the SVG backend:
* `xlink:href` tags' position is specified via `transform=translate(x y)` rather than with `x` and `y` attributes
* `text` tags separate `font` attribute into its separate constituent properties

## PR checklist

- [x] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [x] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- N/A *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/document.html#write-examples-and-tutorials)
- N/A *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/api_changes.html#announce-changes-deprecations-and-new-features)
- [x] Documentation complies with [general](https://matplotlib.org/devdocs/devel/document.html#write-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/document.html#write-docstrings) guidelines

### CODE REVIEW

**Comment by ksunden** [comment]
Did you mean to close this?

I noticed that tests were not passing, but were all failing in the same way, so not sure what the fix is, but it may be relatively easy to do, and we can do it within the same PR.

**Comment by LorenzoPeri17** [comment]
Hi! Thanks for the message. I realized I read about the issue I was having in `lib/matplotlib/testing/compare.py::convert` so I figured I would close the PR because it was failing to do some debugging. 
I have chased down one issue which was described in [this commit](https://github.com/matplotlib/matplotlib/commit/effefafcb95d9140ab1c692b938ba09fdab4719d). I'm more than happy to iron out the remaining issues within this PR if that works best :)

File: lib/matplotlib/testing/compare.py [hunk 1]
```diff
@@ -305,7 +306,16 @@ def convert(filename, cache):
...
             # results), and the tests will fail strangely.
-            if 'style="font:' in contents:
+            if re.search(
+                # searches for attributes :
+                #   style=[font|font-size|font-weight|
+                #          font-family|font-variant|font-style]
+                # taking care of the possibility of multiple style attributes
+                # before the font styling (i.e. opacity)
+                r'style="(?:\w+: [\w \'.]+; )*' +  # unrelated style attributes if any
+                r'font(?:|-size|-weight|-family|-variant|-style):',
```
Comment (timhoffm) [inline lib/matplotlib/testing/compare.py:316]: Can this be simplified to

```suggestion
                r'style="[^"]*font-(size|weight|family|variant|style):',
```
?
Reasoning: We're in the style definition as long as it is not terminated by ". I've also removed the `:?`. While we do not need the group, leaving out the non-matching group qualifier makes this easier to read.

File: lib/matplotlib/tests/test_backend_svg.py [hunk 2]
```diff
@@ -603,12 +603,15 @@ def test_svg_font_string(font_str, include_generic):
...
             map(lambda x: x.strip(), _.strip().split(":"))
             for _ in dict(text_element.items())["style"].split(";")
-        )["font"]
+        )
+        font_size = font_style["font-size"]
+        font_family = font_style["font-family"]
 
-        assert font_info == f"{size}px {font_str}"
+        assert font_size == f"{size}px"
+        assert font_family == font_str
```
Comment (timhoffm) [inline lib/matplotlib/tests/test_backend_svg.py:614]: We don't need the intermediate local variables from l.609/610.
```suggestion
        assert font_style["font-size"] == f"{size}px"
        assert font_style["font-family"] == font_str
```
is more concise.

File: lib/matplotlib/tests/test_backend_svg.py [hunk 3]
```diff
@@ -603,12 +603,15 @@ def test_svg_font_string(font_str, include_generic):
...
     for text_element in tree.findall(f".//{{{ns}}}text"):
         text_count += 1
-        font_info = dict(
+        font_style = dict(
             map(lambda x: x.strip(), _.strip().split(":"))
             for _ in dict(text_element.items())["style"].split(";")
-        )["font"]
+        )
+        font_size = font_style["font-size"]
+        font_family = font_style["font-family"]
```
Comment (timhoffm) [inline lib/matplotlib/tests/test_backend_svg.py:611]:
```suggestion
```

File: lib/matplotlib/tests/test_mathtext.py [hunk 4]
```diff
@@ -453,10 +452,10 @@ def test_mathtext_fallback(fallback, fontlist):
     fig.savefig(buff, format="svg")
     tspans = (ET.fromstring(buff.getvalue())
               .findall(".//{http://www.w3.org/2000/svg}tspan[@style]"))
-    # Getting the last element of the style attrib is a close enough
-    # approximation for parsing the font property.
-    char_fonts = [shlex.split(tspan.attrib["style"])[-1] for tspan in tspans]
-    assert char_fonts == fontlist
+    char_fonts = [
+        re.search(r"font-family: '([\w ]+)';?", tspan.attrib["style"]).group(1)
```
Comment (timhoffm) [inline lib/matplotlib/tests/test_mathtext.py:456]:
```suggestion
        re.search(r"font-family: '([\w ]+)'", tspan.attrib["style"]).group(1)
```
If `;` is optional at the end, we can simply leave it out of the regexp.

File: lib/matplotlib/tests/test_mathtext.py [hunk 5]
```diff
@@ -433,7 +432,7 @@ def test_mathtext_fallback_invalid():
 @pytest.mark.parametrize(
     "fallback,fontlist",
     [("cm", ['DejaVu Sans', 'mpltest', 'STIXGeneral', 'cmr10', 'STIXGeneral']),
-     ("stix", ['DejaVu Sans', 'mpltest', 'STIXGeneral'])])
+     ("stix", ['DejaVu Sans', 'mpltest', 'STIXGeneral', 'STIXGeneral', 'STIXGeneral'])])
```
Comment (timhoffm) [inline lib/matplotlib/tests/test_mathtext.py:435]: Why do we need the additional 'STIXGeneral' entries here?

Comment (LorenzoPeri17) [reply lib/matplotlib/tests/test_mathtext.py:435]: I have slightly changed the logic for how the `tspan`s inside a `text` tag are separated (`backend_svg.py::_draw_text_as_text` line 1224 onwards).
During some testing I found poor compatibility with specifying multiple `x` and `y` in the same `tspan` (from reading [the docs](https://www.w3.org/TR/SVG11/text.html#TSpanElement) maybe `dx` would be better if the old way was to be preferred?)
My new solution is cleaner code-wise and seems to resolve the compatibility issues. However, it causes more `tspans` to be generates, hence why the repetition in the tests.

**Comment by tacaswell** [comment]
To check we need to regenerate the svg files to make keep them in sync with our compare code, not because rendered output (in compliant viewers) actually changed?

**Comment by LorenzoPeri17** [comment]
> To check we need to regenerate the svg files to make keep them in sync with our compare code, not because rendered output (in compliant viewers) actually changed?

Exactly. That and because on the appveyor tests (and only there, azure and github were passung just fine) for some reason a few tests would fail with a very small RMS (~0.001) because of a few pixels in long ticklabels being _slightly_ off when rendering (see attached example)

axvspan_epoch_svg
axvspan_epoch-expected_svg

Falied diff
axvspan_epoch_svg-failed-diff

**Review by timhoffm — APPROVED** [review]

**Comment by tacaswell** [comment]
The appeveyor failures look related and real.

**Comment by LorenzoPeri17** [comment]
I had some time to debug the failing tests. It seems that the baseline images were specifying certain coordinates for `xlink:href` elements beyond 6 digits of precision, causing inkscape to render some text half a pizel off. I have changed this and I can't see any failure.
However, some tests seem to fail now for unrelated reasons. I assume this is because of changes to the workflows. I assume this is because of this fork being behind with respect to `main`?

**Comment by tacaswell** [comment]
Can you please rebase on main and squash all of the image updating into one commit?

**Comment by tacaswell** [comment]
At some point we change from outputting a lot of digits to rounding to a smaller number of digits but did not bulk re-generate the test images because they were still passing.

**Comment by LorenzoPeri17** [comment]
Should be done :)

File: lib/matplotlib/font_manager.py [hunk 6]
```diff
@@ -266,8 +266,11 @@ def _get_fontconfig_fonts():
 @lru_cache
 def _get_macos_fonts():
     """Cache and list the font paths known to ``system_profiler SPFontsDataType``."""
-    d, = plistlib.loads(
-        subprocess.check_output(["system_profiler", "-xml", "SPFontsDataType"]))
+    try:
```
Comment (tacaswell) [inline lib/matplotlib/font_manager.py]: Was this change supposed to be included?

**Comment by tacaswell** [comment]
It looks like there is still a merge from main in here?

File: lib/matplotlib/font_manager.py (hunk 6 above)
Comment (LorenzoPeri17) [reply lib/matplotlib/font_manager.py]: Apologies, that slipped in when squasing. Fixed now

**Comment by LorenzoPeri17** [comment]
Is this OK now?

**Comment by tacaswell** [comment]
no, still a rouge merge from main...is your local main not equal to upstream main?

**Comment by LorenzoPeri17** [comment]
Gotcha! It is done now, I didn't realize I had to rebase, not merge upstream main. I was doing it though the nice green button on the github page, but that only offers the option to do a merge.

**Review by tacaswell — APPROVED** [review]

**Comment by tacaswell** [comment]
Thank you for your work on this @LorenzoPeri17 .

Congratulations on your first merged Matplotlib PR 🎉  I hope we hear from you again!

**Pull request merged**